### PR TITLE
Data Layer: createMiddleware helper

### PR DESCRIPTION
**Pull request by samouri**
_WIP_

testing out an idea for removing the imperativeness / creating a strong pattern for wpcom data-layer requests. It is in a similar vein to  https://github.com/Automattic/wp-calypso/pull/10856.  Results in creating equivalent middleware with only ~60% of the code and they can be fully tested with just input/output assertions and no spies. 

After making a few data-layer middlewares, I noticed a general pattern:

1. using `dispatchRequest` helper util makes life much easier
1. the initiator generates a http action based on the passed in action
2. because the wp-api doesn't fully utilize http status codes, and to make sure calypso doesn't bork, create a `validateResponse()` function to pass things from successHandlers --> error handlers.
3. dispatch actions based on the successful/failure api response.  usually this involves some kind of explicit data shape transformation to be more calypso-friendly.

Because of those steps, the function looks like:

```js
function createMiddleware( {
 	mapActionToHttp,
 	isValidResponse,
 	mapSuccessfulResponseToActions,
 	mapFailureResponseToActions,
 } ) {
```

Theres a big con here:
 - _moar abstraction_
 - what if it only works for a small percent of apis and most need to break out of the pattern?

### CODE REVIEW

File: client/state/data-layer/wpcom/read/tags/mine/delete/test/index.js [hunk 1]
```diff
@@ -29,107 +26,74 @@ const successfulUnfollowResponse = {
...
-			URL: 'https://public-api.wordpress.com/rest/v1/read/tags/poetry/posts'
+			URL: 'https://publicapi.wordpress.com/rest/v1/read/tags/poetry/posts'
 		},
 		{
 			ID: '69750',
 			slug: 'ship',
 			title: 'SHIP',
 			display_name: 'ship',
-			URL: 'https://public-api.wordpress.com/rest/v1/read/tags/ship/posts'
+			URL: 'https://publicapi.wordpress.com/rest/v1/read/tags/ship/posts'
```
Comment (dmsnell) [inline client/state/data-layer/wpcom/read/tags/mine/delete/test/index.js]: are these typos here?

Comment (samouri) [inline client/state/data-layer/wpcom/read/tags/mine/delete/test/index.js]: yessir. not functionally important. will fix

**Comment by dmsnell** [comment]
Here are some high-level initial impressions I have on this PR. Thanks for the continued work @samouri .

Things I like:
 - Prominence of the validator in the function signature
 - Removal of calls to `dispatch` and `next` from within the middleware handlers directly
 - Focus of producing pure actions as a response to network calls

Things I'm hesitant about:
 - Validator but no explicit mapping of data structure from API
 - Can't make stateful requests because `store` is missing from middleware handler signatures (could fix by adding back in)
 - The abstraction seems to add lots of indirection without wrapping much more behavior. If I compare this to just using `dispatchRequest()` it looks like I need new functions that end up wrapping the old functions without new behaviors.
 - It appears like this will introduce the need for including `mergeHandlers` into the middleware files themselves if we listen to more than one action. Not a big problem, but maybe a little awkward in code `mergeHandlers( createMiddleware( ACTION, … ), { [ ACTION ]: thing }, createMiddleware( ACTION, … ) )`

**Comment by samouri** [comment]
Thanks you @dmsnell for the quick feedback! 
Almost all of your comments are easily addressable except this one:

> The abstraction seems to add lots of indirection without wrapping much more behavior. If I compare this to just using dispatchRequest() it looks like I need new functions that end up wrapping the old functions without new behaviors.

This helper _does_ add in more indirection but I also think it makes it easier write/test the code (and helps there be less code a well).  I'm still mulling on whether this abstraction is worth its weight.

---

Now for the easily addressable comments:
> It appears like this will introduce the need for including mergeHandlers into the middleware files themselves if we listen to more than one action. Not a big problem, but maybe a little awkward in code mergeHandlers( createMiddleware( ACTION, … ), { [ ACTION ]: thing }, createMiddleware( ACTION, … ) )

Could change the return.  Currently it is:
```
return {
   [ type ]: [ dispatchRequest( initiator, onSuccess, onError ) ]
};
```
and we could make it:
```
return dispatchRequest( initiator, onSuccess, onError ) ;
```

> Validator but no explicit mapping of data structure from API

I think that `mapSuccessfulResponseToActions` is fairly explicit.  Sure people could just pass through the whole payload without modifying it, but I think this the naming encourages thinking about mapping data from an api to calypso.  If we want to be more explicit I could add in an extra function. (`normalizeResponse` ? ).

**Comment by dmsnell** [comment]
> Could change the return

probably not a good path to follow if a helper function requires that we change one of the fundamental building blocks of the system. @aduth and I had a discussion about that very return type and I felt strongly for keeping it as an object whose keys were Redux action types and whose values were a list of functions to call on those actions.

the structure is there to make it clean and reliable when merging all of the action handlers and the list specifically reminds us that we're adding a function into an existing body of functionality. the current merge strategy is nothing more than taking the union of the lists.

> add in an extra function. (normalizeResponse ? ).

that's a good idea I think. having a first-class piece that says, "hey, where is the normalizer?" when it's missing would send a positive signal to those of us who didn't know it would be good to provide one

File: client/state/data-layer/wpcom/read/tags/mine/delete/test/index.js (hunk 1 above)
Comment (dmsnell) [inline client/state/data-layer/wpcom/read/tags/mine/delete/test/index.js]: let's fix these if for no other reason than to remove them from the PR.

**Comment by dmsnell** [comment]
> it makes it easier write/test the code (and helps there be less code a well)

can you help me understand what about this makes testing easier and how it reduces required code?

_without_ it we end up with independent exportable module-level functions which handle behaviors and can reuse those from other files. _after_ it we end up with functions that only exist at runtime, also it appears like we lose the ability to have stateful middlewares that can sequence multiple dispatches (by returning a list instead of actually dispatching).

if you could further explain what problem this is solving that could help. I'm still not sure I can see what more it's providing us over the existing use of `dispatchRequest`. if the point is to pull in the validator and normalizer then it seems like we could just as easily add those to `dispatchRequest`, but I'm still seeing mostly an empty runtime wrapper around the same old code.

**Comment by samouri** [comment]
> probably not a good path to follow if a helper function requires that we change one of the fundamental building blocks of the system.

I think I may have miscommunicated.  I don't want to change the signature of the data-layer middlewares, I would change what this helper returns.

**Comment by samouri** [comment]
> can you help me understand what about this makes testing easier and how it reduces required code?

Two main things:
1. The biggest thing is how the current data-layer requires you to call dispatch & call next.  That means when writing tests, we need to instantiate spies and verify they were called the right number of times and with the correct arguments.   By instead writing a a series of puret functions, and relying on the framework to call dispatch/next, writing the tests & code becomes a bit slimmer (50 less lines in this case).
2. We need a validator so that we don't need to manually redirect from success handlers --> failure handlers.

> if you could further explain what problem this is solving that could help. I'm still not sure I can see what more it's providing us over the existing use of dispatchRequest

I think this is a good idea 👍.  If we drop this PR we should still add in a mapper + validator to `dispatchRequest`

**Comment by dmsnell** [comment]
> We need a validator so that we don't need to manually redirect from success handlers --> failure handlers.

this is definitely something I think is good and this is a good way to do it (though we would need to pass along the invalid `apiResonse` into the error handler too o it could make appropriate responses or notifications)

> requires you to call dispatch & call next. That means when writing tests, we need to instantiate spies and verify they were called the right number of times and with the correct arguments.

most of our current code is definitely fitting into the pattern of _always dispatch can call next()_ but I wonder how much this is true because of our existing patterns that we haven't yet moved over. _if_ we make the call to `next( action )` implicit I'm worried about misleading developers who don't realize it's there and also of thus passing it along when the intention is to stop the request before further processing. my assumption could be wrong here too.

one concrete example here are polling operations. I think that we'll want the API middleware to throttle certain requests (such as import progress or transfer progress) such that the React components can still send out their requests but this layer can stop those in their tracks if no newer update is needed.

I'm also thinking about related data and being able to take full advantage of the data layer, maybe issuing a second request for something like tag counts after a post update, though maybe that's best handled in the tag middleware listening for post updates…

either way I'm thinking the validator is perfect for the system, maybe even augmenting `dispatchRequest()`.

**Comment by dmsnell** [comment]
@samouri how would you feel about starting fresh on this?

what if we added a final parameter which were the named options, a bucket which could include a validator function and a `fromApi` function…

```js
dispatchRequest( initiator, onSuccess, onFailure, {
  onProgress = noop,
  validator = () => true,
  fromApi = identity,
} )
```

with these options we can run logic like this…

```js
if ( hasData( action ) ) {
	return validator( getData( action ) )
		? onSuccess( store, action, fromApi( getData( action ) ) )
		: onFailure( store, action, { type: VALIDATION_ERROR, error: getData( action ) } )
}

if ( hasError( action ) ) {
	return onFailure( store, action, { type: NETWORK_ERROR, error: getError( action ) } );
}
```

**Comment by samouri** [comment]
@dmsnell: while I still do like the mapping paradigm I introduce here better, its probably worth scrapping so that we can consolidate all of our efforts along one path.

I'll close this PR in favor of introducing the additional helpers like you suggest.

**Comment by samouri** [comment]
Closing as i've created https://github.com/Automattic/wp-calypso/pull/16108

**Pull request closed without merging**